File: chess/pieces/queen.py

from chess import PieceColorEnum
from chess.pieces.default_piece import DefaultPiece
from chess.pieces.piece import BoardSize, Piece, PiecePosition


class Queen(DefaultPiece):
    def __init__(self, color: PieceColorEnum, board_size: BoardSize, position: PiecePosition) -> None:
        super().__init__(color, board_size, position)

    def update_movements(self, board: list[list[Piece | None]]) -> None:
        self.clean_movements()

        directions: list[PiecePosition] = [{'x': 0, 'y': -1},
                                           {'x': 1, 'y': -1},
                                           {'x': 1, 'y': 0},
                                           {'x': 1, 'y': 1},
                                           {'x': 0, 'y': 1},
                                           {'x': -1, 'y': 1},
                                           {'x': -1, 'y': 0},
                                           {'x': -1, 'y': -1}]

        for direction in directions:
            x = self.position['x']
            y = self.position['y']

            while True:
                x = x + direction['x']
                y = y + direction['y']
    
                if not 0 <= x < self.board_size['width'] or not 0 <= y < self.board_size['height']:
                    break

                target = board[y][x]

                if target == None:
                    self.movements[y][x] = True
                    continue

                if target.color != self.color:
                    self.movements[y][x] = True

                break

File: src/main/java/com/vini/app/pieces/Queen.java
package com.vini.app.pieces;

import java.util.List;
import java.util.Objects;

import com.vini.app.fen.FenPieceEnum;
import com.vini.app.types.ColorEnum;

public class Queen extends Piece implements IPiece {
	Queen(ColorEnum color, int[] position, FenPieceEnum fen) {
		super(color, position, fen);
	}

	private final int[][] directions = {
		{-1, -1}, {0, -1}, {1, -1},
		{-1, 0}, {1, 0},
		{-1, 1}, {0, 1}, {1, 1}
	};

	@Override
	public Piece updateMoves(List<List<Piece>> board) {
		for (int[] direction : this.directions) {
			int[] position = this.position;

			while (true) {
				position[0] = position[0] + direction[0];
				position[1] = position[1] + direction[1];

				if (!this.insideBoard(board, position)) {
					break;
				};

				Piece target = board.get(position[1]).get(position[0]);

				if (Objects.isNull(target)) {
					this.moves.get(position[1]).set(position[0], true);
					continue;
				}

				if (!this.isFriend(this, target)) {
					this.moves.get(position[1]).set(position[0], true);
				}

				break;
			}
		}

		return this;
	}
}

